src/devices/cpu/ht1130/ht1130.cpp

```diff
@@ -0,0 +1,1051 @@
+}
+
+void ht1130_device::setcarry()
+{
+	m_carry = 1;
+}
+
+void ht1130_device::clearcarry()
+{
+	m_carry = 0;
+}
+
+void ht1130_device::settimer(u8 data)
+{
+	m_timer = data;
+}
+
+void ht1130_device::settimer_upper(u8 data)
+{
+	m_timer = (m_timer & 0xf) | (data & 0xf) << 4;
+}
+
+void ht1130_device::settimer_lower(u8 data)
+{
+	m_timer = (m_timer & 0xf0) | (data & 0xf);
+}
+
+u8 ht1130_device::gettimer_upper()
+{
+	return (m_timer >> 4) & 0xf;
+}
+
+u8 ht1130_device::gettimer_lower()
+{
+	return m_timer & 0xf;
+}
+
+u8 ht1130_device::getr1r0()
+{
+	return (getreg(1)<<4) | getreg(0);
+}
+
+u8 ht1130_device::getr1r0_data()
+{
+	const u8 dataaddress = getr1r0();
+	return space(AS_DATA).read_byte(dataaddress);
+}
+
+void ht1130_device::setr1r0_data(u8 data)
+{
+	const u8 dataaddress = getr1r0();
+	space(AS_DATA).write_byte(dataaddress, data);
+}
+
+u8 ht1130_device::getr3r2()
+{
+	return (getreg(3)<<4) | getreg(2);
+}
+
+u8 ht1130_device::getr3r2_data()
+{
+	const u8 dataaddress = getr3r2();
+	return space(AS_DATA).read_byte(dataaddress);
+}
+
+void ht1130_device::setr3r2_data(u8 data)
+{
+	const u8 dataaddress = getr3r2();
+	space(AS_DATA).write_byte(dataaddress, data);
+}
+
+
+void ht1130_device::internal_data_map(address_map &map)
+{
+	map(0x00, 0x7f).ram().w(FUNC(ht1130_device::tempram_w)).share("tempram");
+	map(0xe0, 0xff).ram().w(FUNC(ht1130_device::displayram_w)).share("displayram");
+}
+
+void ht1190_device::internal_data_map_ht1190(address_map &map)
+{
+	map(0x00, 0x9f).ram().w(FUNC(ht1190_device::tempram_w)).share("tempram");
+	map(0xb0, 0xff).ram().w(FUNC(ht1190_device::displayram_w)).share("displayram");
+}
+
+void ht1130_device::device_start()
+{
+	space(AS_PROGRAM).cache(m_cache);
+	space(AS_PROGRAM).specific(m_space);
+
+	set_icountptr(m_icount);
+
+	state_add(HT1130_PC, "PC", m_pc);
+	state_add(STATE_GENPC, "GENPC", m_pc).noshow();
+	state_add(STATE_GENPCBASE, "CURPC", m_pc).noshow();
+
+	state_add(HT1130_ACC, "ACC", m_acc);
+
+	for (int i = 0; i < 5; i++)
+		state_add(HT1130_R0 + i, string_format("R%d", i).c_str(), m_regs[i]);
+
+	state_add(HT1130_TIMER_EN, "TIMER_EN", m_timer_en);
+	state_add(HT1130_TIMER, "TIMER", m_timer);
+
+
+	save_item(NAME(m_pc));
+	save_item(NAME(m_icount));
+
+	save_item(NAME(m_regs));
+	save_item(NAME(m_acc));
+	save_item(NAME(m_carry));
+	save_item(NAME(m_irqen));
+	save_item(NAME(m_timer_en));
+	save_item(NAME(m_inhalt));
+	save_item(NAME(m_timerover));
+	save_item(NAME(m_timer));
+	save_item(NAME(m_stackaddr));
+	save_item(NAME(m_stackcarry));
+}
+
+void ht1130_device::device_reset()
+{
+	m_pc = 0;
+
+	m_regs[0] = m_regs[1] = m_regs[2] = m_regs[3] = m_regs[4] = 0;
+	m_acc = 0;
+	m_carry = 0;
+	m_irqen = 0;
+	m_timer_en = 0;
+	m_inhalt = 0;
+	m_timerover = 0;
+	m_timer = 0;
+	m_stackaddr = 0;
+	m_stackcarry = 0;
+}
+
+u8 ht1130_device::fetch()
+{
+	m_icount--; // single byte opcodes take a single cycle, double take 2
+	if (m_timer_en)
+		m_timer++;
+	return m_space.read_byte(m_pc++);
+}
```

Comment:
Since you’ve got instructions that take multiple cycles and you’re making no provisions for returning to the scheduler midway through the instruction, you need to override `execute_max_cycles` and return the maximum number of cycles an instruction can take.